app/services/district_report_service.rb

```diff
@@ -0,0 +1,91 @@
+class DistrictReportService
+  def initialize(facilities:, selected_date:)
+    @facilities = Array(facilities)
+    @selected_date = selected_date
+    @data = {
+      controlled_patients: {},
+      registrations: {},
+      cumulative_registrations: 0,
+      quarterly_registrations: []
+    }.with_indifferent_access
+  end
+
+  attr_reader :selected_date, :facilities, :data
+
+  def call
+    compile_control_and_registration_data
+
+    compile_cohort_trend_data
+
+    data
+  end
+
+  def compile_control_and_registration_data
+    beginning_of_period = selected_date.advance(months: -11)
+    previous_registrations = 0
+    registrations.each do |(period, count)|
```

Comment:
This a bit complicated to understand, from what I gather from the logic, could this be simply achieved by the following:

```
previous_registrations =
  PatientRegistrationsPerDayPerFacility
    .where(facility: @facilities)
    .group(:year, :month)
    .where('year < ?', beginning_of_period.year)
    .or(PatientRegistrationsPerDayPerFacility
          .where('month <= ?', beginning_of_period.month))
    .sum(:registration_count)
```